File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.ts
function shipWithinDays(weights: number[], days: number): number {
    let l = Math.max(...weights);
    let r = weights.reduce((a,b) => a+b, 0);
    while(l < r) {
        let m = Math.floor((r-l)/2 + l);
        let cur = 0, need = 1;
        for(const w of weights) {
            if(cur + w > m){
                ++need;
                cur = 0;
            }
            cur += w;
        }
        if(need > days) l = m + 1;
        else r = m;
    }
    return l;
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int left = 0, right = 0;
        for(int w : weights) {
            left = Math.max(left, w);
            right += w;
        }
        while(left < right) {
            int mid = (right-left)/2 + left;
            int cur = 0, need = 1;
            for(int w : weights) {
                if(cur + w > mid) {
                    cur = 0;
                    need++;
                }
                cur += w;
            }
            if(need > days) left = mid + 1;
            else right = mid;
        }
        return left;
    }
}